### DISCUSSION TITLE
scoped_session kwargs and join_transaction_mode

### DISCUSSION THREAD

**Discussion by auxsvr** [sqlalchemy/sqlalchemy#10598] [Usage Questions]
It seems that the new parameter `join_transaction_mode` is in the spirit of https://github.com/sqlalchemy/sqlalchemy/discussions/8610. The idea is that I use async scoped sessions to assign a connection to a task, then call the scoped session specifying the `join_transaction_mode` in each function inside the task, so that I avoid explicit use of the session as an argument in each called function, I avoid complicated transaction management logic and everything works as expected. Currently, this fails raising `InvalidRequestError`, as new arguments are not allowed for an existing session. I wonder if this should be reconsidered, as `join_transaction_mode` is specifically designed to allow for transaction reuse.

`join_transaction_mode='create_savepoint'` could then be used for sessions in utility functions, so that the entire function savepoint is rolled back or committed depending on local state and the calling code can also roll back or commit the entire transaction depending on the function's result.

Sample code:
```python
db_session = async_scoped_session(sessionmaker(create_async_engine('url')), scopefunc=asyncio.current_task)

async def test():
    async with db_session(join_transaction_mode='create_savepoint') as session:
        c = await session.scalar(select(models.D).limit(1))
        return c

async with db_session() as session:
    d = await test()
````
throws 

> InvalidRequestError: Scoped session is already present; no new arguments may be specified.

**Comment by CaselIT** [comment]
Hi,

> Currently, this fails raising InvalidRequestError, as new arguments are not allowed for an existing session

Could you provide an example of what is raising that issue? Thanks

**Reply by auxsvr** [reply]
Hi. Thank you for the quick response, I've updated the question.

**Comment by CaselIT** [comment]
Thanks for sharing the example.
What you are trying to do will not work, since the session object returned by a call to db_session is always the same for a particular task.
`join_transaction_mode` is used when you create a new session from an existing connection that's already in a transaction.

also from your description it seems you can just do
```py
async with db_session.begin_nested():
  await db_session.execute(...)
```

**Reply by auxsvr** [reply]
Your solution seems to work, but I was hoping to avoid subtransactions unless absolutely necessary because of [the performance impact on PostgreSQL](https://about.gitlab.com/blog/2021/09/29/why-we-spent-the-last-month-eliminating-postgresql-subtransactions/). For example, if the function starts with the session in a transaction, use a subtransaction, otherwise do as `rollback_only`, which is the behaviour of `conditional_savepoint`, or even `control_fully`, depending on requirements.

Now that I think about it more, my request may make `begin_nested` redundant. Wouldn't this be ideal, as most (all?) of the use cases for subtransactions are covered by `join_transaction_mode`?

**Reply by CaselIT** [reply]
I suggested begin_nested since your example had `create_savepoint` that will use savepoints.

I'm not sure what you are trying to achieve here. Could better explain what you want to do?

If you just want to share the same session in the same task without passing it around then the `async_scoped_session` already does it, just use execute and similar methods without need of an explicit begin

**Reply by auxsvr** [reply]
The idea is to nest transactions or reuse the existing transaction in functions, avoiding savepoints if possible. The logic in `join_transaction_mode` seems to cover this requirement, but currently it applies only to transactions started in another session. The request is to make it applicable for transactions in the same session, effectively making `begin_nested` redundant, but simplifying the code users write.

**Reply by auxsvr** [reply]
Do you think that a PR for this has any chance to be accepted? As I am not familiar with SA internals, I will probably need some help with this; any advice is welcome.

**Reply by zzzeek** [reply]
it looks like you are attempting to restore some version of SQLAlchemy's now -removed behavior of allowing subtransactions.

The join_transcation_mode thing does not take any effect unless the session is bound to a single connection that's already running in a tranasction, so you should adapt your code to work that way.  it's made for test harnesses.
